distributed/deploy/cluster.py

```diff
@@ -70,8 +89,22 @@ async def _start(self):
         self._watch_worker_status_task = asyncio.ensure_future(
             self._watch_worker_status(comm)
         )
+
+        with contextlib.suppress(KeyError):
+            self.cluster_info.update(
+                (await self.scheduler_comm.get_metadata(keys=["cluster-manager-info"]))
+            )
+        self._sync_cluster_info_task = asyncio.ensure_future(self._sync_cluster_info())
```

Comment:
This is likely leaving an asyncio Task behind after test teardown, isn't it? Instead of the `future + while True` you can use a tornado periodic callback which handles this properly. the PC should be stopped when the cluster is closed, though.

Reply:
I started with a PC but was struggling with the async stuff. The coroutine never seems to get called.

Do you have any suggestions/

Reply:
I have added some cancellation code, but a PC would be preferable.

Reply:
PC not being called is often connected to the wrong event loop being active. I'm note entirely sure but often `pytest-asyncio` is messing with our event loop infrastructure.

You could try if instead of using the `pytest.mark.asyncio` our old-shool `gen_test` decorator helps.

Reply:
If you add the PC to the attribute `periodic_callbacks` the start routine _should_ take care of properly starting the PC such that it is attached to the same loop. How one can enforce this (ugly, but should work) is something like

```python
from tornado.ioloop import IOLoop
self.periodic_callbacks['sync-cluster'] = self.pc = PeriodicCallback(self._sync_cluster_info)
loop = IOLoop.get_current()
loop.add_callback(self.pc.start)
```

Reply:
Switching to `gen_test` gives this error.

```
E       fixture 'func' not found
>       available fixtures: a, anyio_backend, anyio_backend_name, anyio_backend_options, b, c, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, cleanup, client, client_contract_script, client_secondary, cluster_fixture, cov, doctest_namespace, event_loop, invalid_python_script, kind_cluster, loop, loop_in_thread, monkeypatch, no_cover, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, s, security, testrun_uid, tls_client, tls_cluster, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, valid_python_script, worker_id, zmq_ctx
>       use 'pytest --fixtures [testpath]' for help on them.
```

Reply:
This is how I've implemented PC but the test fails because the PC never seems to start.

```diff
diff --git a/distributed/deploy/cluster.py b/distributed/deploy/cluster.py
index 747a9c9c..cde9760f 100644
--- a/distributed/deploy/cluster.py
+++ b/distributed/deploy/cluster.py
@@ -8,7 +8,7 @@ import uuid
 from contextlib import suppress
 from inspect import isawaitable
 
-from tornado.ioloop import PeriodicCallback
+from tornado.ioloop import PeriodicCallback, IOLoop
 
 import dask.config
 from dask.utils import _deprecated, format_bytes, parse_timedelta
@@ -61,7 +61,6 @@ class Cluster:
         self._asynchronous = asynchronous
         self._watch_worker_status_comm = None
         self._watch_worker_status_task = None
-        self._sync_cluster_info_task = None
         self._cluster_manager_logs = []
         self.quiet = quiet
         self.scheduler_comm = None
@@ -94,7 +93,11 @@ class Cluster:
             self.cluster_info.update(
                 (await self.scheduler_comm.get_metadata(keys=["cluster-manager-info"]))
             )
-        self._sync_cluster_info_task = asyncio.ensure_future(self._sync_cluster_info())
+
+        self.periodic_callbacks["sync-cluster-info"] = pc = PeriodicCallback(
+            self._sync_cluster_info, 1000
+        )
+        IOLoop.current().add_callback(pc.start)
         self.status = Status.running
 
     async def _sync_cluster_info(self):
@@ -119,8 +122,6 @@ class Cluster:
             await self._watch_worker_status_comm.close()
         if self._watch_worker_status_task:
             await self._watch_worker_status_task
-        if self._sync_cluster_info_task:
-            self._sync_cluster_info_task.cancel()
 
         for pc in self.periodic_callbacks.values():
             pc.stop()
```

Reply:
It should work but the first iteration is only scheduled rather late. Below is a diff where I added this interval to be added as an option. this way the test completes in ~0.2s and the PC works as expected. If you are fine with it, I can push this as a commit

```diff
commit 276ae107126e3450eb7c4e2e45ca1bda3baae3a3
Author: fjetter <fjetter@users.noreply.github.com>
Date:   Thu Jul 8 12:25:11 2021 +0200

    Use PC instead of future

diff --git a/distributed/deploy/cluster.py b/distributed/deploy/cluster.py
index 747a9c9c..5587c79b 100644
--- a/distributed/deploy/cluster.py
+++ b/distributed/deploy/cluster.py
@@ -8,7 +8,7 @@ import uuid
 from contextlib import suppress
 from inspect import isawaitable
 
-from tornado.ioloop import PeriodicCallback
+from tornado.ioloop import IOLoop, PeriodicCallback
 
 import dask.config
 from dask.utils import _deprecated, format_bytes, parse_timedelta
@@ -55,7 +55,7 @@ class Cluster:
     _supports_scaling = True
     name = None
 
-    def __init__(self, asynchronous, quiet=False, name=None):
+    def __init__(self, asynchronous, quiet=False, name=None, scheduler_sync_interval=1):
         self.scheduler_info = {"workers": {}}
         self.periodic_callbacks = {}
         self._asynchronous = asynchronous
@@ -66,6 +66,9 @@ class Cluster:
         self.quiet = quiet
         self.scheduler_comm = None
         self._adaptive = None
+        self._sync_interval = parse_timedelta(
+            scheduler_sync_interval, default="seconds"
+        )
 
         if name is None:
             name = str(uuid.uuid4())[:8]
@@ -94,19 +97,18 @@ class Cluster:
             self.cluster_info.update(
                 (await self.scheduler_comm.get_metadata(keys=["cluster-manager-info"]))
             )
-        self._sync_cluster_info_task = asyncio.ensure_future(self._sync_cluster_info())
+        self.loop = IOLoop.current()
+        self.periodic_callbacks["sync-cluster-info"] = pc = PeriodicCallback(
+            self._sync_cluster_info, self._sync_interval * 1000
+        )
+        self.loop.add_callback(pc.start)
         self.status = Status.running
 
     async def _sync_cluster_info(self):
-        try:
-            while True:
-                await self.scheduler_comm.set_metadata(
-                    keys=["cluster-manager-info"],
-                    value=copy.copy(self.cluster_info),
-                )
-                await asyncio.sleep(1)
-        except asyncio.CancelledError:
-            return
+        await self.scheduler_comm.set_metadata(
+            keys=["cluster-manager-info"],
+            value=copy.copy(self.cluster_info),
+        )
 
     async def _close(self):
         if self.status == Status.closed:
@@ -119,15 +121,13 @@ class Cluster:
             await self._watch_worker_status_comm.close()
         if self._watch_worker_status_task:
             await self._watch_worker_status_task
-        if self._sync_cluster_info_task:
-            self._sync_cluster_info_task.cancel()
-
-        for pc in self.periodic_callbacks.values():
-            pc.stop()
 
         if self.scheduler_comm:
             await self.scheduler_comm.close_rpc()
 
+        for pc in self.periodic_callbacks.values():
+            pc.stop()
+
         self.status = Status.closed
 
     def close(self, timeout=None):
diff --git a/distributed/deploy/local.py b/distributed/deploy/local.py
index e759d8b1..d5fc94e9 100644
--- a/distributed/deploy/local.py
+++ b/distributed/deploy/local.py
@@ -120,6 +120,7 @@ class LocalCluster(SpecCluster):
         interface=None,
         worker_class=None,
         scheduler_kwargs=None,
+        scheduler_sync_interval=1,
         **worker_kwargs,
     ):
         if ip is not None:
@@ -240,6 +241,7 @@ class LocalCluster(SpecCluster):
             asynchronous=asynchronous,
             silence_logs=silence_logs,
             security=security,
+            scheduler_sync_interval=scheduler_sync_interval,
         )
 
     def start_worker(self, *args, **kwargs):
diff --git a/distributed/deploy/spec.py b/distributed/deploy/spec.py
index aa3dc3f7..33ef1c84 100644
--- a/distributed/deploy/spec.py
+++ b/distributed/deploy/spec.py
@@ -241,6 +241,7 @@ class SpecCluster(Cluster):
         silence_logs=False,
         name=None,
         shutdown_on_close=True,
+        scheduler_sync_interval=1,
     ):
         self._created = weakref.WeakSet()
 
@@ -270,6 +271,7 @@ class SpecCluster(Cluster):
         super().__init__(
             asynchronous=asynchronous,
             name=name,
+            scheduler_sync_interval=scheduler_sync_interval,
         )
 
         if not self.asynchronous:
diff --git a/distributed/deploy/tests/test_local.py b/distributed/deploy/tests/test_local.py
index 930d5156..20392239 100644
--- a/distributed/deploy/tests/test_local.py
+++ b/distributed/deploy/tests/test_local.py
@@ -1083,13 +1083,16 @@ async def test_local_cluster_redundant_kwarg(nanny):
 
 @pytest.mark.asyncio
 async def test_cluster_info_sync():
-    async with LocalCluster(processes=False, asynchronous=True) as cluster:
+    async with LocalCluster(
+        processes=False, asynchronous=True, scheduler_sync_interval="2ms"
+    ) as cluster:
         assert cluster.cluster_info["name"] == cluster.name
+        await asyncio.sleep(0.005)
         info = await cluster.scheduler_comm.get_metadata(keys=["cluster-manager-info"])
         assert info["name"] == cluster.name
         info = cluster.scheduler.get_metadata(keys=["cluster-manager-info"])
         assert info["name"] == cluster.name
 
         cluster.cluster_info["foo"] = "bar"
-        await asyncio.sleep(2)
+        await asyncio.sleep(0.005)
         assert "foo" in cluster.scheduler.get_metadata(keys=["cluster-manager-info"])

```

Reply:
> If you are fine with it, I can push this as a commit

@fjetter yes please!